### PR TITLE
SAK-33855 allow students to view assignment/course grade stats

### PR DESCRIPTION
**Pull request by steveswinsburg**
Adds a new setting to Gradebook that allows an instructor to control the visibility of stats to students.
Includes some boilerplate deletion -> replace with lombok.
Fixes some code that was missing braces / should be final / replace deprecated Wicket methods etc.

Note that a database upgrade is required and this has already been merged into master via:
https://github.com/sakaiproject/sakai-reference/pull/44

### CODE REVIEW

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java [hunk 1]
```diff
@@ -23,165 +23,105 @@
 import org.apache.commons.lang3.builder.ToStringBuilder;
 import org.apache.commons.lang3.builder.ToStringStyle;
 
+import lombok.Getter;
```
Comment (ern) [inline edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java]: Looks like you might be able to use @Data here as this simplifies it even more!

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java [hunk 2]
```diff
@@ -23,6 +23,9 @@
 import org.apache.commons.lang3.builder.HashCodeBuilder;
 import org.apache.commons.lang3.builder.ToStringBuilder;
 
+import lombok.Getter;
```
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java]: Same maybe lombok @Data is a better fit here?

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java [hunk 3]
```diff
@@ -70,314 +156,35 @@ public Gradebook() {
...
-        append("id", id).
-        append("uid", uid).
-        append("name", name).toString();
+        append("id", this.id).
+        append("uid", this.uid).
+        append("name", this.name).toString();
     }
 
     @Override
-	public boolean equals(Object other) {
```
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java]: Lombok as @EqualsAndHashcode to make this simpler, you can control which fields should be used in the comparison.
For example to add an equals and hashcode on just the uid you could use:
`@EqualsAndHashcode(of = {"uid"})`

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java [hunk 4]
```diff
@@ -118,7 +120,9 @@ public void addGradebook(final String uid, final String name) {
             // Check for null default.
             if (defaultGradeMapping == null) {
                 defaultGradeMapping = (GradeMapping)gradeMappings.iterator().next();
-                if (log.isWarnEnabled()) log.warn("No default GradeMapping found for new Gradebook=" + gradebook.getUid() + "; will set default to " + defaultGradeMapping.getName());
+                if (log.isWarnEnabled()) {
+					log.warn("No default GradeMapping found for new Gradebook=" + gradebook.getUid() + "; will set default to " + defaultGradeMapping.getName());
```
Comment (ern) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: Should use newer slf4j logging methods:
`log.warn("No default GradeMapping found for new Gradebook={}; will set default to {}", gradebook.getUid(), defaultGradeMapping.getName());`

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java [hunk 5]
```diff
@@ -217,48 +227,51 @@ public void saveGradeMappingToGradebook(final String scaleUuid, final String gra
...
 	}
 
-	private void mergeGradeMappings(Collection gradingScaleDefinitions, Session session) throws HibernateException {
-		Map newMappingDefinitionsMap = new HashMap();
-		HashSet uidsToSet = new HashSet();
-		for (Iterator iter = gradingScaleDefinitions.iterator(); iter.hasNext(); ) {
-			GradingScaleDefinition bean = (GradingScaleDefinition)iter.next();
+	private void mergeGradeMappings(final Collection gradingScaleDefinitions, final Session session) throws HibernateException {
+		final Map newMappingDefinitionsMap = new HashMap();
+		final HashSet uidsToSet = new HashSet();
```
Comment (ern) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: add generics to HashSet

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java [hunk 6]
```diff
@@ -169,14 +176,17 @@ private List addDefaultGradingScales(Session session) throws HibernateException
 			gradingScale.setGrades(new ArrayList(sampleMapping.getGrades()));
 			gradingScale.setDefaultBottomPercents(new HashMap(sampleMapping.getGradeMap()));
 			session.save(gradingScale);
-			if (log.isInfoEnabled()) log.info("Added Grade Mapping " + gradingScale.getUid());
+			if (log.isInfoEnabled()) {
+				log.info("Added Grade Mapping " + gradingScale.getUid());
```
Comment (ern) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: update new logging call

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java [hunk 7]
```diff
@@ -217,48 +227,51 @@ public void saveGradeMappingToGradebook(final String scaleUuid, final String gra
             // templates, so if for some reason none have been defined yet,
             // do that now.
             if (gradingScales.isEmpty()) {
-                if (log.isInfoEnabled()) log.info("No Grading Scale defined yet. This is probably because you have upgraded or you are working with a new database. Default grading scales will be created. Any customized system-wide grade mappings you may have defined in previous versions will have to be reconfigured.");
+                if (log.isInfoEnabled()) {
+					log.info("No Grading Scale defined yet. This is probably because you have upgraded or you are working with a new database. Default grading scales will be created. Any customized system-wide grade mappings you may have defined in previous versions will have to be reconfigured.");
```
Comment (ern) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: update logging call

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java [hunk 8]
```diff
@@ -298,15 +315,20 @@ private void mergeGradeMappings(Collection gradingScaleDefinitions, Session sess
 			GradingScaleDefinition bean = (GradingScaleDefinition)newMappingDefinitionsMap.get(uid);
 			copyDefinitionToScale(bean, gradingScale);
 			session.save(gradingScale);
-			if (log.isInfoEnabled()) log.info("Added Grading Scale " + gradingScale.getUid());
+			if (log.isInfoEnabled()) {
+				log.info("Added Grading Scale " + gradingScale.getUid());
```
Comment (ern) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: update logging call

**Comment by steveswinsburg** [comment]
Hi @ern switching this over to use @Data. I haven't been fond of this in the past due to this issue:
https://github.com/rzwitserloot/lombok/issues/345

But let's give it another shot.

In terms of the SecurityAdvisor, I wanted to take this approach so that the entire Gradebook service didn't need to be modified. There are a number of permission checks sprinkled all over the place in there and removing them all might open up security issues elsewhere. I basically need a student to be able to to retrieve grades for everyone in the class, so made this security intrusion as small as possible.

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java (hunk 3 above)
Comment (steveswinsburg) [reply edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java]: Done. FYI `of` is being deprecated so new pattern is:
````
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@ToString(onlyExplicitlyIncluded = true)
public class SomeClass {
...
```
Then on the field:
````
@ToString.Include
@EqualsAndHashCode.Include
private String field;
````
👍

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java (hunk 2 above)
Comment (steveswinsburg) [reply edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/Gradebook.java]: Done

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java (hunk 5 above)
Comment (steveswinsburg) [reply edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java]: Done as park of a cleanup of that entire class

**Comment by steveswinsburg** [comment]
@ern are we good to go here?

**Pull request merged**